aidial_adapter_vertexai/chat/tools.py

```diff
@@ -0,0 +1,209 @@
+        function_call: Literal["auto", "none", "required"] | FunctionChoice,
+        functions: List[Function],
+    ) -> Tuple[bool, List[Function]]:
+        match function_call:
+            case "none":
+                return False, []
+            case "auto":
+                return False, functions
+            case "required":
+                if not functions:
+                    raise ValidationError("No functions are available")
+                return True, functions
+            case FunctionChoice(name=name):
+                new_functions = [
+                    func for func in functions if func.name == name
+                ]
+                if not new_functions:
+                    raise ValidationError(
+                        f"Function {name!r} is not on the list of available functions"
+                    )
+                return True, new_functions
+            case _:
+                assert_never(function_call)
+
+    @staticmethod
+    def tool_choice_to_function_call(
+        tool_choice: Literal["auto", "none", "required"] | ToolChoice | None,
+    ) -> Literal["auto", "none", "required"] | FunctionChoice | None:
+        match tool_choice:
+            case ToolChoice(function=FunctionChoice(name=name)):
+                return FunctionChoice(name=name)
+            case _:
+                return tool_choice
+
+    @classmethod
+    def from_request(cls, request: AzureChatCompletionRequest) -> Self:
+        validate_messages(request)
+
+        if request.functions is not None:
+            functions = request.functions
+            function_call = request.function_call
+            tool_ids = None
+
+        elif request.tools is not None:
+            functions = [tool.function for tool in request.tools]
+            function_call = ToolsConfig.tool_choice_to_function_call(
+                request.tool_choice
+            )
+            tool_ids = collect_tool_ids(request.messages)
+
+        else:
+            functions = []
+            function_call = None
+            tool_ids = None
+
+        if function_call is None:
+            function_call = "auto" if functions else "none"
+
+        required, selected = ToolsConfig.filter_functions(
+            function_call, functions
+        )
+
+        return cls(functions=selected, required=required, tool_ids=tool_ids)
+
+    def to_gemini_tools(self) -> List[GeminiTool] | None:
+        if not self.functions:
+            return None
+
+        return [
+            GeminiTool(
+                function_declarations=[
+                    GeminiFunction(
+                        name=func.name,
+                        parameters=func.parameters,
+                        description=func.description,
+                    )
+                    for func in self.functions
+                ]
+            )
+        ]
+
+    def to_gemini_tool_config(self) -> GeminiToolConfig | None:
+        if not self.functions or not self.required:
```

Comment:
Shouldn't the "required" property control the mode: ANY vs AUTO?

Reply:
Good catch. Fixed.